Mathlib/MeasureTheory/Function/LpSeminorm/NNLpNorm.lean

```diff
@@ -0,0 +1,200 @@
+
+lemma nnLpNorm_nsmul [NormedSpace ℝ E] (n : ℕ) (f : α → E) (μ : Measure α) :
+    nnLpNorm (n • f) p μ = n • nnLpNorm f p μ := by simp [nnLpNorm, eLpNorm_nsmul]
+
+variable [NormedSpace ℝ 𝕜]
+
+lemma nnLpNorm_natCast_mul (n : ℕ) (f : α → 𝕜) (p : ℝ≥0∞) (μ : Measure α) :
+    nnLpNorm ((n : α → 𝕜) * f) p μ = n * nnLpNorm f p μ := by
+  simpa only [nsmul_eq_mul] using nnLpNorm_nsmul n f μ
+
+lemma nnLpNorm_natCast_mul' (n : ℕ) (f : α → 𝕜) (p : ℝ≥0∞) (μ : Measure α) :
+    nnLpNorm (n * f ·) p μ = n * nnLpNorm f p μ := nnLpNorm_natCast_mul ..
+
+lemma nnLpNorm_mul_natCast (f : α → 𝕜) (n : ℕ) (p : ℝ≥0∞) (μ : Measure α) :
+    nnLpNorm (f * (n : α → 𝕜)) p μ = nnLpNorm f p μ * n := by
+  simpa only [mul_comm] using nnLpNorm_natCast_mul n f p μ
+
+lemma nnLpNorm_mul_natCast' (f : α → 𝕜) (n : ℕ) (p : ℝ≥0∞) (μ : Measure α) :
+    nnLpNorm (f · * n) p μ = nnLpNorm f p μ * n := nnLpNorm_mul_natCast ..
+
+lemma nnLpNorm_div_natCast [CharZero 𝕜] {n : ℕ} (hn : n ≠ 0) (f : α → 𝕜) (p : ℝ≥0∞)
+    (μ : Measure α) : nnLpNorm (f / (n : α → 𝕜)) p μ = nnLpNorm f p μ / n := by
+  rw [eq_div_iff (by positivity), ← nnLpNorm_mul_natCast]; simp [Pi.mul_def, hn]
+
+lemma nnLpNorm_div_natCast' [CharZero 𝕜] {n : ℕ} (hn : n ≠ 0) (f : α → 𝕜) (p : ℝ≥0∞)
+    (μ : Measure α) : nnLpNorm (f · / n) p μ = nnLpNorm f p μ / n := nnLpNorm_div_natCast hn ..
+
+end NormedField
+
+lemma nnLpNorm_add_le (hf : MemLp f p μ) (hg : MemLp g p μ) (hp : 1 ≤ p) :
+    nnLpNorm (f + g) p μ ≤ nnLpNorm f p μ + nnLpNorm g p μ := by
+  unfold nnLpNorm
+  rw [← ENNReal.toNNReal_add hf.eLpNorm_ne_top hg.eLpNorm_ne_top]
+  gcongr
+  exacts [ENNReal.add_ne_top.2 ⟨hf.eLpNorm_ne_top, hg.eLpNorm_ne_top⟩,
+    eLpNorm_add_le hf.aestronglyMeasurable hg.aestronglyMeasurable hp]
+
+lemma nnLpNorm_sub_le (hf : MemLp f p μ) (hg : MemLp g p μ) (hp : 1 ≤ p) :
+    nnLpNorm (f - g) p μ ≤ nnLpNorm f p μ + nnLpNorm g p μ := by
+  simpa [sub_eq_add_neg] using nnLpNorm_add_le hf hg.neg hp
+
+lemma nnLpNorm_le_nnLpNorm_add_nnLpNorm_sub' (hf : MemLp f p μ) (hg : MemLp g p μ) (hp : 1 ≤ p) :
+    nnLpNorm f p μ ≤ nnLpNorm g p μ + nnLpNorm (f - g) p μ := by
+  simpa using nnLpNorm_add_le hg (hf.sub hg) hp
+
+lemma nnLpNorm_le_nnLpNorm_add_nnLpNorm_sub (hf : MemLp f p μ) (hg : MemLp g p μ) (hp : 1 ≤ p) :
+    nnLpNorm f p μ ≤ nnLpNorm g p μ + nnLpNorm (g - f) p μ := by
+  simpa [neg_add_eq_sub] using nnLpNorm_add_le hg.neg (hg.sub hf) hp
+
+lemma nnLpNorm_le_add_nnLpNorm_add (hf : MemLp f p μ) (hg : MemLp g p μ) (hp : 1 ≤ p) :
+    nnLpNorm f p μ ≤ nnLpNorm (f + g) p μ + nnLpNorm g p μ := by
+  simpa using nnLpNorm_add_le (hf.add hg) hg.neg hp
+
+lemma nnLpNorm_sub_le_nnLpNorm_sub_add_nnLpNorm_sub (hf : MemLp f p μ) (hg : MemLp g p μ)
+    (hh : MemLp h p μ) (hp : 1 ≤ p) :
+    nnLpNorm (f - h) p μ ≤ nnLpNorm (f - g) p μ + nnLpNorm (g - h) p μ := by
+  simpa using nnLpNorm_add_le (hf.sub hg) (hg.sub hh) hp
+
+lemma nnLpNorm_sum_le {ι : Type*} {s : Finset ι} {f : ι → α → E} (hf : ∀ i ∈ s, MemLp (f i) p μ)
```

Comment:
Do we also have the eLpNorm of this and the below? (Do we want them?)

Reply:
`eLpNorm_sum_le` exists. `eLpNorm_expect_le` can't be stated (because `Module ℚ≥0 ℝ≥0∞` currently doesn't exist). I'll leave a TODO

Reply:
Perhaps you'd also like to mention it here, so that once that does exist it can be used to golf this one?

Reply:
I think the current proof of `nnLpNorm_expect_le` is longer than the one from the hypothetical `eLpNorm_expect_le `: Look at how long the proof of `nnLpNorm_sum_le` is.

Reply:
You can make it shorter:
```
  rw [← ENNReal.coe_le_coe, coe_nnLpNorm_eq_eLpNorm (memLp_finset_sum' s hf),
    ENNReal.coe_finset_sum]
  exact (eLpNorm_sum_le (fun i hi ↦ (hf _ hi).aestronglyMeasurable) hp).trans_eq <|
    Finset.sum_congr rfl fun i hi ↦ (coe_nnLpNorm_eq_eLpNorm (hf i hi)).symm
```